dm/dm/ctl/master/start_stop_validation.go

```diff
@@ -33,105 +34,149 @@ const (
-		RunE:  startStopValidation(StartValidationOp),
+		RunE:  startValidation,
 	}
 	cmd.Flags().Bool("all-task", false, "whether applied to all tasks")
-	cmd.Flags().String("mode", "full", "specify the mode of validation: full, fast")
+	cmd.Flags().String("mode", "", "specify the mode of validation: full, fast")
+	cmd.Flags().String("start-time", "", "specify the start time of binlog for validation, e.g. '2021-10-21 00:01:00' or 2021-10-21T00:01:00")
 	return cmd
 }
 
 func NewStopValidationCmd() *cobra.Command {
 	cmd := &cobra.Command{
 		Use:   "stop [-s source ...] [--all-task] [task-name]",
 		Short: "stop validating the completeness of the data",
-		RunE:  startStopValidation(StopValidationOp),
+		RunE:  stopValidation,
 	}
 	cmd.Flags().Bool("all-task", false, "whether to all tasks")
 	return cmd
 }
 
-func startStopValidation(op string) func(*cobra.Command, []string) error {
-	formatStartStopValidationError := func(cmd *cobra.Command, errMsg string) error {
-		cmd.SetOut(os.Stdout)
-		common.PrintCmdUsage(cmd)
-		return errors.New(errMsg)
+type validationStartStopArgs struct {
+	sources  []string
+	allTask  bool
+	taskName string
+
+	mode      string
+	startTime string
+	// whether user has set --mode or --start-time explicitly
+	explicitMode      bool
+	explicitStartTime bool
+}
+
+func printUsageAndFailWithMessage(cmd *cobra.Command, errMsg string) error {
+	cmd.SetOut(os.Stdout)
+	common.PrintCmdUsage(cmd)
+	return errors.New(errMsg)
+}
+
+func parseValidationStartStopArgs(cmd *cobra.Command, op string) (validationStartStopArgs, string, bool) {
+	var err error
+	args := validationStartStopArgs{}
+	if args.sources, err = common.GetSourceArgs(cmd); err != nil {
+		return args, err.Error(), false
 	}
-	return func(cmd *cobra.Command, _ []string) error {
-		var (
-			sources   []string
-			isAllTask bool
-			taskName  string
-			err       error
-		)
-		sources, err = common.GetSourceArgs(cmd)
-		if err != nil {
-			return err
-		}
-		isAllTask, err = cmd.Flags().GetBool("all-task")
-		if err != nil {
-			return err
-		}
-		switch len(cmd.Flags().Args()) {
-		case 1:
-			taskName = cmd.Flags().Arg(0)
-			if isAllTask {
-				// contradiction
-				return formatStartStopValidationError(cmd, "either `task-name` or `all-task` should be set")
-			}
-		case 0:
-			if !isAllTask {
-				// contradiction
-				return formatStartStopValidationError(cmd, "either `task-name` or `all-task` should be set")
-			}
-		default:
-			return formatStartStopValidationError(cmd, "too many arguments are specified")
-		}
-		ctx, cancel := context.WithCancel(context.Background())
-		defer cancel()
-		if op == StartValidationOp {
-			// TODO: get `from-time` flag
-			var mode string
-			mode, err = cmd.Flags().GetString("mode")
+	if args.allTask, err = cmd.Flags().GetBool("all-task"); err != nil {
+		return args, err.Error(), false
+	}
+
+	if op == StartValidationOp {
+		args.explicitMode = cmd.Flags().Changed("mode")
```

Comment:
since the default value of `mode` is "", we in fact don't need this check. Also for `start-time`.

I'm both OK with these two styles: if explicitMode {checkMode} or if mode != "" {checkMode}

Reply:
will leave it.